src/main/java/io/synadia/flink/source/NATSJetstreamSource.java

```diff
@@ -0,0 +1,98 @@
+package io.synadia.flink.source;
+
+import io.nats.client.Connection;
+import io.synadia.flink.Utils;
+import io.synadia.flink.common.ConnectionFactory;
+import io.synadia.flink.source.enumerator.NatsSourceEnumerator;
+import io.synadia.flink.source.split.NatsSubjectCheckpointSerializer;
+import io.synadia.flink.source.split.NatsSubjectSplit;
+import io.synadia.flink.source.split.NatsSubjectSplitSerializer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.connector.source.Boundedness;
+import org.apache.flink.api.connector.source.Source;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NATSJetstreamSource<T> implements Source<T, NatsSubjectSplit, Collection<NatsSubjectSplit>>, ResultTypeQueryable<T> {
+
+    private final ConnectionFactory connectionFactory;
+    private final String natsSubject;
+    private final DeserializationSchema<T> deserializationSchema;
+    private final NATSConsumerConfig config;
+    private static final Logger LOG = LoggerFactory.getLogger(NATSJetstreamSource.class);
+    private final String id;
+
+    // Package-private constructor to ensure usage of the Builder for object creation
+    NATSJetstreamSource(DeserializationSchema<T> deserializationSchema, ConnectionFactory connectionFactory, String natsSubject, NATSConsumerConfig config) {
```

Comment:
Flink has inbuilt implementation for Deserializer and Serializers.
In our application code, we are heavily invested in using Flink's native Deserializer (& Serializers) as our current Pulsar-Flink connector implements it.  In my humble opinion, this would be the same for other folks moving from another messaging framework. Thoughts?

Reply:
I don't think so. Consider the PayloadDeserializer interface
```
OutputT getObject(String subject, byte[] input, Headers headers);
```
versus the primary methods in the DeserializationSchema
```
T deserialize(byte[] var1) throws IOException;
```
The Nats message is more than just the data byte array. But there is your code could implement both interfaces and delegate the getObject implementation to call deserialize.